datafusion-examples/examples/memory_pool_tracking.rs

```diff
@@ -0,0 +1,191 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This example demonstrates how to use TrackConsumersPool for memory tracking and debugging.
+//!
+//! The TrackConsumersPool provides enhanced error messages that show the top memory consumers
+//! when memory allocation fails, making it easier to debug memory issues in DataFusion queries.
+//!
+//! # Examples
+//!
+//! * [`automatic_usage_example`]: Shows how to use RuntimeEnvBuilder to automatically enable memory tracking
+//! * [`manual_tracking_example`]: Shows how to manually create and configure a TrackConsumersPool
+
+use datafusion::execution::memory_pool::{
+    GreedyMemoryPool, MemoryConsumer, MemoryPool, TrackConsumersPool,
+};
+use datafusion::execution::runtime_env::RuntimeEnvBuilder;
+use datafusion::prelude::*;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    println!("=== DataFusion Memory Pool Tracking Example ===\n");
+
+    // Example 1: Automatic Usage with RuntimeEnvBuilder
+    automatic_usage_example().await?;
+
+    println!("\n{}\n", "=".repeat(60));
+
+    // Example 2: Manual tracking with custom consumers
+    manual_tracking_example()?;
+
+    println!("\n{}\n", "=".repeat(60));
+
+    Ok(())
+}
+
+/// Example 1: Automatic Usage with RuntimeEnvBuilder
+///
+/// This shows the recommended way to use TrackConsumersPool through RuntimeEnvBuilder,
+/// which automatically creates a TrackConsumersPool with sensible defaults.
+async fn automatic_usage_example() -> datafusion::error::Result<()> {
+    println!("Example 1: Automatic Usage with RuntimeEnvBuilder");
+    println!("------------------------------------------------");
+
+    // Success case: Create a runtime with reasonable memory limit
+    println!("Success case: Normal operation with sufficient memory");
+    let runtime = RuntimeEnvBuilder::new()
+        .with_memory_limit(5_000_000, 1.0) // 5MB, 100% utilization
+        .build_arc()?;
+
+    let config = SessionConfig::new();
+    let ctx = SessionContext::new_with_config_rt(config, runtime);
+
+    // Create a simple table for demonstration
+    ctx.sql("CREATE TABLE test AS VALUES (1, 'a'), (2, 'b'), (3, 'c')")
+        .await?
+        .collect()
+        .await?;
+
+    println!("✓ Created table with memory tracking enabled");
+
+    // Run a simple query to show it works
+    let results = ctx.sql("SELECT * FROM test").await?.collect().await?;
+    println!(
+        "✓ Query executed successfully. Found {} rows",
+        results.len()
+    );
+
+    println!("\n{}", "-".repeat(50));
+
+    // Error case: Create a runtime with low memory limit to trigger errors
+    println!("Error case: Triggering memory limit error with detailed error messages");
+
+    // Use a WITH query that generates data and then processes it to trigger memory usage
```

Comment:
I think this will be a nice base to start from when implementing more sophisticated memory reporting / tracking